src/Corax/Queries/MemoizationMatch.Provider.cs

```diff
@@ -9,7 +9,7 @@
 namespace Corax.Queries
 {
 
-    public unsafe struct MemoizationMatchProvider<TInner> : IMemoizationMatchSource
+    public unsafe class MemoizationMatchProvider<TInner> : IMemoizationMatchSource
```

Comment:
I want to explain myself here.

I decided to change this into a class to make it more convenient to pass between multiple objects etc. Please notice this contains states so it requires passing it as reference (and this is not so easy to handle and there is a risk to lose reference somewhere and corrupt the data inside the buffer).

Reply:
No, don't do it like this. As a general direction in cases where such a thing is necessary for whatever reason; do the following:

Introduce this class. 
```csharp
public unsafe class MemoizationMatchProviderRef<TInner> :  IMemoizationMatchSource where TInner ....
{
       private MemoizationMatchProvider<TInner> _inner;
} 
```
and implement every single method with aggressive inlining as a pass-through. This is essentially a custom boxing but it will solve for you many things among those not losing the optimizations path at the JIT.